### PR TITLE
Fix dedicated io transform leaving unused shapes

### PR DESCRIPTION
**Pull request by milesziemer**
This commit changes the way `CreateDedicatedInputOutput` transform checks for singular references to a shape being used as input or output in an operation. It now checks to make sure the shape is only referenced by the specified operation. This fixes an edge case where a shape was being used as both the input and output for an operation, and would be left unused in the model after the transformation was applied.  Two new test cases were added to verify the behavior when a shape is used as both input and output.

*Issue #, if available:* #1373

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: smithy-model/src/test/resources/software/amazon/smithy/model/transform/dedicated-input-output/handles-dangling-single-reference-before.smithy [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+// We have a shape being used by both input and output
+// so there will be an extra or "dangling" reference
+// to `GetFooOutput` from the input
+operation GetFoo {
+    input: GetFooOutput
+    output: GetFooOutput
+}
+
+structure GetFooOutput {}
```
Comment (srchase) [inline smithy-model/src/test/resources/software/amazon/smithy/model/transform/dedicated-input-output/handles-dangling-single-reference-before.smithy:13]: Can the naming of this output structure get updated so it's clear that it will be removed in the `-after.smithy` file?

File: smithy-model/src/test/resources/software/amazon/smithy/model/transform/dedicated-input-output/handles-dangling-single-reference-before.smithy [hunk 2]
```diff
@@ -0,0 +1,13 @@
+$version: "2.0"
+
+namespace smithy.example
+
+// We have a shape being used by both input and output
+// so there will be an extra or "dangling" reference
```
Comment (srchase) [inline smithy-model/src/test/resources/software/amazon/smithy/model/transform/dedicated-input-output/handles-dangling-single-reference-before.smithy:6]: nit: This comment is a little confusing without the context of the PR and original issue. Maybe it can just describe what this test is showing, that a shared input/output shape gets cleaned up and replaced by the appropriate input and output shapes.

File: smithy-model/src/main/java/software/amazon/smithy/model/transform/CreateDedicatedInputAndOutput.java [hunk 3]
```diff
@@ -153,36 +156,22 @@ private StructureShape renameShapeIfNeeded(
...
-            if (rel.getRelationshipType().getDirection() == RelationshipDirection.DIRECTED) {
-                totalDirectedEdges++;
-                if (rel.getRelationshipType() != expected) {
-                    return false;
-                }
-            }
-        }
-
-        return totalDirectedEdges == 1;
+        // edges to the expected shape
```
Comment (srchase) [inline smithy-model/src/main/java/software/amazon/smithy/model/transform/CreateDedicatedInputAndOutput.java:171]: nit: missing period.

File: smithy-model/src/test/resources/software/amazon/smithy/model/transform/dedicated-input-output/handles-dangling-single-reference-before.smithy (hunk 1 above)
Comment (milesziemer) [reply smithy-model/src/test/resources/software/amazon/smithy/model/transform/dedicated-input-output/handles-dangling-single-reference-before.smithy]: I added a second test that verifies that the unused shape will be removed, I'll leave this open to make sure that looks ok

**Review by srchase — APPROVED** [review]
